parboiled-core/src/main/scala-3/org/parboiled2/support/TailSwitch.scala

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2009-2019 Mathias Doenitz
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.parboiled2.support
+
+import hlist._
+
+import scala.annotation.implicitNotFound
+
+/**
+  * type-level implementation of this logic:
+  *   Out =
+  *     R                      if T has a tail of type L
+  *     (L dropRight T) ::: R  if L has a tail of type T
+  */
+@implicitNotFound("Illegal rule composition")
+sealed trait TailSwitch[L <: HList, T <: HList, R <: HList] {
+  type Out <: HList
+}
+object TailSwitch {
+  type Reverse0[Acc <: HList, L <: HList] <: HList = L match {
```

Comment:
In the case, when it loops, it does so in `Reverse0` but I cannot see any obvious reason for that.